### PR TITLE
Implements logic for Product Lookup

### PR DESCRIPTION
**Pull request by sharadkjaiswal**
- Defines interfaces IInstalledProduct, IProductLookUp and
IProductCollection to implement product lookup logic.

- InstalledProductLookUp implements basic lookup algorithm to get
product installations using registry keys

- DynamoProducts implements logic to find all Dynamo installations.

- Fixes http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-6703

**Reviewers:**
- [ ] @ikeough 
- [x] @aparajit-pratap

### CODE REVIEW

**Comment by sharadkjaiswal** [comment]
It seems I can't reference DynamoUtilities from version selector dll, so I have to move ProductLookUp classes and intefraces to DynamoRevitVersionSelector project.

File: src/Tools/DynamoInstallDetective/ProductLookUp.cs [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+        public override bool Equals(object obj)
+        {
+            return CompareTo(obj) == 0;
+        }
+
+        public override int GetHashCode()
+        {
+            return VersionInfo.GetHashCode();
+        }
+    }
```
Comment (aparajit-pratap) [inline src/Tools/DynamoInstallDetective/ProductLookUp.cs]: I don't think you need to override `GetHashCode` and `Equals` if all you're doing is implementing `IComparable` for sorting.

**Comment by aparajit-pratap** [comment]
LGTM

**Comment by sharadkjaiswal** [comment]
Thanks @aparajit-pratap

**Pull request merged**